### ISSUE TITLE
System.BadImageFormatException using tasks in F# 6.0

### ISSUE THREAD

**Issue by Lanayx** [dotnet/fsharp#12189]
Please provide a succinct description of the issue.

**Repro steps**

The [following test](https://github.com/fsharplang-ru/pulsar-client-dotnet/blob/fsharpsix/tests/IntegrationTests/Reader.fs#L277-L279) fails in [Pulsar.Client repo, fsharpsix branch](https://github.com/fsharplang-ru/pulsar-client-dotnet/tree/fsharpsix) when built with new msbuild
(**Microsoft (R) Build Engine version 17.0.0-preview-21460-01+8f208e609 for .NET**)
in runtime ([link to failed github action](https://github.com/fsharplang-ru/pulsar-client-dotnet/runs/3708635451?check_suite_focus=true#step:12:256))

```
System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
   at <StartupCode$Pulsar-Client>.$ConsumerImpl.clo@122-84.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Pulsar.Client.Api.ConsumerImpl`1.wrapPostAndReply[U](Task`1 mbAsyncReply) in D:\Work\Pulsar.Client\src\Pulsar.Client\Internal\ConsumerImpl.fs:line 122
   at Pulsar.Client.Api.ConsumerImpl`1.Pulsar.Client.Api.IConsumer<'T>.SeekAsync(MessageId messageId) in D:\Work\Pulsar.Client\src\Pulsar.Client\Internal\ConsumerImpl.fs:line 1569
   at Pulsar.Client.Internal.ReaderImpl`1.Pulsar.Client.Api.IReader<'T>.SeekAsync(MessageId msgId) in D:\Work\Pulsar.Client\src\Pulsar.Client\Internal\ReaderImpl.fs:line 79
   at Pulsar.Client.IntegrationTests.Reader.basicReaderCheck@35-1.MoveNext() in D:\Work\Pulsar.Client\tests\IntegrationTests\Reader.fs:line 61
```

which is failing at the start of the following function

```fsharp
    let wrapPostAndReply (mbAsyncReply: Task<ResultOrException<'A>>) =
        task {
            match! mbAsyncReply with
            | Ok msg ->
                return msg
            | Error ex ->
                return reraize ex
        }
```

When building the solution with old compiler
( **Microsoft (R) Build Engine version 16.10.2+857e5a733 for .NET Framework** )
I'm getting build errors, they [ can be observed in github actions](https://github.com/fsharplang-ru/pulsar-client-dotnet/runs/3708310234?check_suite_focus=true) 
![image](https://user-images.githubusercontent.com/3329606/134773354-d07ee3d0-9ac0-4ac5-be89-8780c8317cdc.png)

I suspect this might be the issue

Unfortunately I couldn't reproduce it on a simpler case for now, maybe later or someone else will help

Steps to run integration tests:
1. Start Pulsar cluster (assuming Docker desktop is installed)
```
git clone https://github.com/fsharplang-ru/pulsar-client-dotnet
git checkout fsharpsix
cd tests\compose\standalone
docker-compose up
```
Wait about 2 minutes until the messages flow in console stops (you'll see the message 'Transaction coordinator metadata setup success')

2. Run integration tests  
```
dotnet build
cd tests\IntegrationTests
dotnet run
```

**Expected behavior**

The test passes successully when build with both old and new compiler, i.e. it should just work like TaskBuilder.fs worked. (develop branch in Pulsar.Client repo)

**Actual behavior**

When build with new compiler tests fail in runtime
When build with old compiler code doesn't compile

**Known workarounds**

Not to upgrade to 6.0

**Related information**

* Windows 10
* Library is built as .NET Standard 2.0
* Jetbrains Rider

**Comment by dsyme** [comment]
@gsomix Could you write the repro steps?  Also so I need OSX/Ubuntu to repro or can I do this on Windows?

I tried the following on Windows so far
```
git clone https://github.com/fsharplang-ru/pulsar-client-dotnet
cd pulsar-client-dotnet
dotnet tool restore
dotnet restore
dotnet build
git checkout fsharpsix
dotnet restore
dotnet build
cd tests\UnitTests
dotnet test
dotnet test -c Release
```

and no problems so far

**Comment by gsomix** [comment]
@dsyme Do you mean @Lanayx? It seems integration tests fail, not unit ones. I tried to make minimal repro, but not succeed.

**Comment by dsyme** [comment]
> @dsyme Do you mean @Lanayx? It seems integration tests fail, not unit ones. I tried to make minimal repro, but not succeed.

Yes, thanks

**Comment by Lanayx** [comment]
@dsyme I couldn't create a repro with simple case :( To run integration tests you need to setup pulsar cluster by going to the `tests\compose\standalone\` folder and running `docker-compose up` (assuming you have docker desktop installed)

**Comment by dsyme** [comment]
Perhaps an exception is being raised elsewhere, then caught, then re-raised, and the problem is not with `wrapPostAndReply `

**Comment by Lanayx** [comment]
No the execution doesn't reach there. I've modified the method to be

![image](https://user-images.githubusercontent.com/3329606/134950883-7edd290f-8307-4e82-a669-e55f117ce44f.png)

And it fails with the same exception, if I replace task builder with Task.FromResult as in commented code it works

**Comment by Lanayx** [comment]
Just made another observation, if I change the method it to

```fsharp
    let wrapPostAndReply (mbAsyncReply: Task<ResultOrException<unit>>) =
        task {
            match! mbAsyncReply with
            | Ok msg ->
                return msg
            | Error ex ->
                return reraize ex
        }
```
e.g. 'A replaced with unit then there is no error

**Comment by gsomix** [comment]
@dsyme @Lanayx OK, first successful attempt of minimal repro.

```fsharp
open System
open System.Runtime.ExceptionServices
open System.Threading.Tasks

let reraize<'a> ex =
    (ExceptionDispatchInfo.Capture ex).Throw()
    Unchecked.defaultof<'a>

type internal ResultOrException<'T> = Result<'T, exn>

type ConsumerImpl<'T>() =
    let wrapPostAndReply (mbAsyncReply: Task<ResultOrException<'A>>) =
        task {
            match! mbAsyncReply with
            | Ok msg ->
                return msg
            | Error ex ->
                return reraize ex
        }

    member this.SeekAsync() =
        task { return Ok () } 
        |> wrapPostAndReply 

[<EntryPoint>]
let main args = 
    let consumer = ConsumerImpl()
    consumer.SeekAsync() |> ignore
    0
```

**Comment by gsomix** [comment]
And even smaller repro. Please note the both generic parameters `'T` and `'A` are crucial for `BadImageFormatException`.

```fsharp
open System.Threading.Tasks

type Issue12189<'T>() =
    let run (_: Task<'A>) =
        task { return Unchecked.defaultof<'A> }

    member _.Run() =
        run (Task.FromResult(42))

[<EntryPoint>]
let main args = 
    let tmp = Issue12189()
    tmp.Run() |> ignore
    0
```

**Comment by dsyme** [comment]
Thank yo, I can see the fix that's needed

**Comment by dsyme** [comment]
Fixed by #12195
